### ISSUE TITLE
Wow, another TopDrawer person!

### ISSUE THREAD

**Issue by celane** [cepgen/topdrawer#1]
I was just about to see about starting a TopDrawer repo to preserve the code....The version I have was copied from a site in Japan, which got it from Bonner Lab (Texas?) which got it from FNAL, which got it from SLAC. 

The 'exponential fit' needs some work, but oh is that old Ratfor code horrible.

**Comment by forthommel** [comment]
Indeed, and you were fast to spot it, as it is only pushed since a couple of days now. I did not even take the time to write a proper README to mention [the obscure source](https://web.archive.org/web/20170208154458/https://home.fnal.gov/~parke/TD/) of this version.
If for some reason you have a more modern version, I would be very happy either to store it here, or to fork your future repository :slightly_smiling_face: 

As a matter of fact, I am still trying to find the one I used for the development of [the CepGen TopDrawer wrapper](https://github.com/cepgen/cepgen/blob/master/CepGenAddOns/TopdrawerWrapper/TopdrawerDrawer.cpp) which could digest a piped set of commands as an argument, instead of the filename of an intermediate .top file.

**Comment by celane** [comment]
Ah, I know Stephen Parke. But that webpage only has executables, not
source. I was going to post the source (horribly ugly as it is, it can
serve as an example of how NOT to program). 

doing "strings ./td | grep Version", I see that Stephen has a Bonner Lab
version of 13 Apr 1995; the copy I have has Bonner Lab 9 Aug 1995. But
Stephen's version has CERN histogram code (RZ etc.) which mine doesn't,
although mine still has some super-antique "manhist" histogram stuff,
IIRC, so I guess there was some parallel development back in the day.

Do you have source for the Parke version? I made some fixes to the
version I have, but it would be good to merge into a version with the
most features and the fewest bugs. 




On Tue, 17 Oct 2023 16:40:07 -0700
Laurent Forthomme ***@***.***> wrote:

> Indeed, and you were fast to spot it, as it is only pushed since a
> couple of days now. I did not even take the time to write a proper
> README to mention [the obscure
> source](https://web.archive.org/web/20170208154458/https://home.fnal.gov/~parke/TD/)
> of this version. If for some reason you have a more modern version, I
> would be very happy either to store it here, or to fork your future
> repository :slightly_smiling_face: 
> 
> As a matter of fact, I am still trying to find the one I used for the
> development of [the CepGen TopDrawer
> wrapper](https://github.com/cepgen/cepgen/blob/master/CepGenAddOns/TopdrawerWrapper/TopdrawerDrawer.cpp)
> which could digest a piped set of commands as an argument, instead of
> the filename of an intermediate .top file.
> 



-- 
Drexel University  |/             /    ·   ν ·   Chuck Lane   |D
======]－－－>－－－C－－－－－π+－－<                Disque 911  |U
Particle Physics    \             \    ~~ e+~~   215-895-1545 |N
***@***.***                    －​μ+－－+νν                 |E

**Comment by forthommel** [comment]
Hello @celane and apologies for the late reply. Maybe an email thread can be more efficient 😀

I unfortunately do not have any version of the source compilable on a modern architecture, but I can surely have a look at yours if publicly available!

**Comment by celane** [comment]
No problem with the delay; I've been using all my spare cycles
getting pnp4nagios fixed up (an online plotting utility for
Nagios monitoring). Near the end of that effort now, so I can
pay attention to other things. 

"any version of the source compilable on a modern architecture"

Heh. Well, the source I have is MORTRAN, based on FORTRAN IV
(I think? Doesn't seem very Fortran 77 ish).
So yeah, old. But I have gotten it to compile. Takes a bit of 
beating on gfortran compiler options, but it does compile and
run.  

The reason I ask is that the executable you posted claims a 
version of 5.13, while the version I have is 5.12, although
the "date" on the version I have is slightly later.  And 
it looks like the 5.13 version has some more modern
CERN histogram code (HBOOK 4 vs HBOOK 3).  It does look
like the update to 5.13 took out some useful user
interface code, so I think it would make sense to start 
with the 5.12 version and port in the more recent histogram
routines, since that's probably easier. So any source code 
for 5.13 is potentially useful. 

If messing "under the hood" of Topdrawer, it would be really nice to
update the code to allow fetching data from ROOT files. And 
there's some fitting routines that seem not to be working very
well and could use some debugging.  


On Thu, 23 Nov 2023 15:54:24 -0800
Laurent Forthomme ***@***.***> wrote:

> Hello @celane and apologies for the late reply. Maybe an email thread
> can be more efficient 😀
> 
> I unfortunately do not have any version of the source compilable on a
> modern architecture, but I can surely have a look at yours if
> publicly available!
> 



-- 
Drexel University  |/             /    ·   ν ·   Chuck Lane   |D
======]－－－>－－－C－－－－－π+－－<                Disque 911  |U
Particle Physics    \             \    ~~ e+~~   215-895-1545 |N
***@***.***                    －​μ+－－+νν                 |E

**Comment by forthommel** [comment]
Thanks for your report on the reviving process of this little piece of history, which made me discover the whole `MORTRAN` universe! For sure, anything can be considered in interfacing more modern I/O toolchains, provided the user base is interested in integrating it in their environment!
I would surely be interested in having a look at your "`5.12+`", and possibly playing a bit with the automation of its build on some common architectures!
